File: MiniDOMWebCtrls/DMD/Forms/Charts/CChartValue.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Forms


Namespace Forms

    <Serializable> _
    Public Class CChartValue
        Implements IComparable, XML.IDMDXMLSerializable

        Private m_Label As String
        Private m_X As Single
        Private m_Value As Single?
        Private m_Color As System.Drawing.Color
        <NonSerialized> Private m_Serie As CChartSerie
        Private m_RenderColor As System.Drawing.Color = Drawing.Color.Empty

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            Me.m_Label = ""
            Me.m_X = 0
            Me.m_Value = Nothing
            Me.m_Color = Drawing.Color.Empty
            Me.m_Serie = Nothing
        End Sub

        ''' <summary>
        ''' Restituisce il grafico a cui appartiene questo oggetto
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Chart As CChart
            Get
                If (Me.m_Serie) Is Nothing Then Return Nothing
                Return Me.m_Serie.Chart
            End Get
        End Property

        ''' <summary>
        ''' Restituisce la serie a cui appartiene questo oggetto
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Serie As CChartSerie
            Get
                Return Me.m_Serie
            End Get
        End Property
        Protected Friend Sub SetSerie(ByVal value As CChartSerie)
            Me.m_Serie = value
        End Sub

        ''' <summary>
        ''' Restituisce il valore dell'etichetta visualizzata in corrispondenza di questo elemento
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Label As String
            Get
                Return Me.m_Label
            End Get
            Set(value As String)
                Me.m_Label = value
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta il valore dell'elemento (ordinata)
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Value As Single?
            Get
                Return Me.m_Value
            End Get
            Set(value As Single?)
                If (value = Me.m_Value) Then Exit Property
                Me.m_Value = value
                Me.InvalidateChart()
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta la posizione dell'oggetto (ascissa)
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property X As Single
            Get
                Return Me.m_X
            End Get
            Set(value As Single)
                If (Me.m_X = value) Then Exit Property
                Me.m_X = value
                Me.InvalidateChart()
            End Set
        End Property

        Public Property Color As System.Drawing.Color
            Get
                Return Me.m_Color
            End Get
            Set(value As System.Drawing.Color)
                Me.m_Color = value
            End Set
        End Property

        Private Function CompareTo(obj As CChartValue) As Integer
            Return Arrays.Compare(Me.m_Value, obj.m_Value)
        End Function

        Private Function CompareTo1(obj As Object) As Integer Implements IComparable.CompareTo
            Return Me.CompareTo(obj)
        End Function

        Protected Sub InvalidateChart()
            If (Me.Chart IsNot Nothing) Then Me.Chart.Invalidate()
        End Sub

        Public Function GetRenderColor() As System.Drawing.Color
            If Me.Color.Equals(System.Drawing.Color.Empty) Then
                If (Me.m_RenderColor.Equals(System.Drawing.Color.Empty)) Then
                    Dim i As Integer = Me.Serie.Values.IndexOf(Me)
                    If (i >= 0 AndAlso i <= UBound(CChart.DefaultColors)) Then
                        Me.m_RenderColor = Colors.FromWeb(CChart.DefaultColors(i))
                    Else
                        Me.m_RenderColor = Utils.FormsUtils.GetRandomColor
                    End If
                End If
                Return Me.m_RenderColor
            Else
                Return Me.Color
            End If
        End Function


        Protected Overridable Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "Label" : Me.m_Label = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "X" : Me.m_X = XML.Utils.Serializer.DeserializeDouble(fieldValue)
                Case "Value" : Me.m_Value = XML.Utils.Serializer.DeserializeDouble(fieldValue)
                Case "Color" : Me.m_Color = Utils.FormsUtils.FromRGBSTR(XML.Utils.Serializer.DeserializeString(fieldValue))
                Case "RenderColor", Me.m_RenderColor = Utils.FormsUtils.FromRGBSTR(XML.Utils.Serializer.DeserializeString(fieldValue))
            End Select
        End Sub

        Protected Overridable Sub XMLSerialize(writer As XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("Label", Me.m_Label)
            writer.WriteAttribute("X", Me.m_X)
            writer.WriteAttribute("Value", Me.m_Value)
            writer.WriteAttribute("Color", Utils.FormsUtils.ToRGBSTR(Me.m_Color))
            writer.WriteAttribute("RenderColor", Utils.FormsUtils.ToRGBSTR(Me.m_RenderColor))
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class


End Namespace

File: MiniDOMWebCtrls/DMD/Forms/Charts/CChartValue.cs
using System;
using DMD;
using DMD.XML;

namespace minidom.Forms
{
    [Serializable]
    public class CChartValue 
        : IComparable, IDMDXMLSerializable
    {
        private string m_Label;
        private float m_X;
        private float? m_Value;
        private System.Drawing.Color m_Color;
        [NonSerialized]
        private CChartSerie m_Serie;
        private System.Drawing.Color m_RenderColor = System.Drawing.Color.Empty;

        public CChartValue()
        {
            DMDObject.IncreaseCounter(this);
            m_Label = "";
            m_X = 0f;
            m_Value = default;
            m_Color = System.Drawing.Color.Empty;
            m_Serie = null;
        }

        /// <summary>
        /// Restituisce il grafico a cui appartiene questo oggetto
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public CChart Chart
        {
            get
            {
                if (m_Serie is null)
                    return null;
                return m_Serie.Chart;
            }
        }

        /// <summary>
        /// Restituisce la serie a cui appartiene questo oggetto
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public CChartSerie Serie
        {
            get
            {
                return m_Serie;
            }
        }

        protected internal void SetSerie(CChartSerie value)
        {
            m_Serie = value;
        }

        /// <summary>
        /// Restituisce il valore dell'etichetta visualizzata in corrispondenza di questo elemento
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public string Label
        {
            get
            {
                return m_Label;
            }

            set
            {
                m_Label = value;
            }
        }

        /// <summary>
        /// Restituisce o imposta il valore dell'elemento (ordinata)
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public float? Value
        {
            get
            {
                return m_Value;
            }

            set
            {
                if (value == m_Value == true)
                    return;
                m_Value = value;
                InvalidateChart();
            }
        }

        /// <summary>
        /// Restituisce o imposta la posizione dell'oggetto (ascissa)
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public float X
        {
            get
            {
                return m_X;
            }

            set
            {
                if (m_X == value)
                    return;
                m_X = value;
                InvalidateChart();
            }
        }

        public System.Drawing.Color Color
        {
            get
            {
                return m_Color;
            }

            set
            {
                m_Color = value;
            }
        }

        private int CompareTo(CChartValue obj)
        {
            return DMD.Arrays.Compare(m_Value, obj.m_Value);
        }

        int IComparable.CompareTo(object obj)
        {
            return CompareTo((CChartValue)obj);
        }

        protected void InvalidateChart()
        {
            if (Chart is object)
                Chart.Invalidate();
        }

        public System.Drawing.Color GetRenderColor()
        {
            if (Color.Equals(System.Drawing.Color.Empty))
            {
                if (m_RenderColor.Equals(System.Drawing.Color.Empty))
                {
                    int i = Serie.Values.IndexOf(this);
                    if (i >= 0 && i <= DMD.Arrays.UBound(CChart.DefaultColors))
                    {
                        m_RenderColor = Colors.FromWeb(CChart.DefaultColors[i]);
                    }
                    else
                    {
                        m_RenderColor = Utils.FormsUtils.GetRandomColor();
                    }
                }

                return m_RenderColor;
            }
            else
            {
                return Color;
            }
        }

        protected virtual void SetFieldInternal(string fieldName, object fieldValue)
        {
            switch (fieldName ?? "")
            {
                case "Label":
                    {
                        m_Label = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "X":
                    {
                        m_X = (float)DMD.XML.Utils.Serializer.DeserializeDouble(DMD.Strings.CStr(fieldValue));
                        break;
                    }

                case "Value":
                    {
                        m_Value = (float?)DMD.XML.Utils.Serializer.DeserializeDouble(DMD.Strings.CStr(fieldValue));
                        break;
                    }

                case "Color":
                    {
                        m_Color = Utils.FormsUtils.FromRGBSTR(DMD.XML.Utils.Serializer.DeserializeString(fieldValue));
                        break;
                    }

                case "RenderColor":
                    this.m_RenderColor = Utils.FormsUtils.FromRGBSTR(DMD.XML.Utils.Serializer.DeserializeString(fieldValue));
                    break;
                     
            }
        }

        protected virtual void XMLSerialize(XMLWriter writer)
        {
            writer.WriteAttribute("Label", m_Label);
            writer.WriteAttribute("X", m_X);
            writer.WriteAttribute("Value", m_Value);
            writer.WriteAttribute("Color", Utils.FormsUtils.ToRGBSTR(m_Color));
            writer.WriteAttribute("RenderColor", Utils.FormsUtils.ToRGBSTR(m_RenderColor));
        }

        ~CChartValue()
        {
            DMDObject.DecreaseCounter(this);
        }

        void IDMDXMLSerializable.XMLSerialize(XMLWriter writer)
        {
            this.XMLSerialize(writer);
        }

        void IDMDXMLSerializable.SetFieldInternal(string fieldName, object fieldValue)
        {
            this.SetFieldInternal(fieldName, fieldValue);
        }

    }
}